### DISCUSSION TITLE
[TV] Generate inputs from a list or asset

### DISCUSSION THREAD

**Discussion by Glujaz** [HomeSpan/HomeSpan#817] [General]
Hi there (again, same noob).

Different topic here. I'm trying to clean up a lot my code, and was starting to use a "for loop" for this line when adding a new input source :

    `SpanService *hdmi1 = new Service::InputSource();`

So, I started to do something like this :

  ```
for (int i = 1; i = inputCount; i++) {
    SpanService *inputList[i].index= new Service::InputSource();
    new Characteristic::ConfiguredName(inputList[i].name,true);
    new Characteristic::Identifier(i,true);
    new Characteristic::IsConfigured(1;true);            // Source included in the Settings Screen...
    new Characteristic::CurrentVisibilityState(1,true);  // ...and included in the Selection List...
    new Characteristic::TargetVisibilityState(0,true);   // ...and a "checked" checkbox is provided on the Settings Screen that can be used to toggle CurrentVisibilityState()
  }
```

With this being how I made the Array and list :

```
struct Commands {
  String index;
  String name;
  String set;
  String get;
};

Commands inputList[] = {

////////////Input List///////////
  {"input1","CD", "cd", "cd"},
  {"input2","Coaxial 1", "coax1", "coax1"},
  {"input3","Coaxial 2", "coax2", "coax2"},
  //////////////////////////////////
};
const int inputCount = sizeof(inputList) / sizeof(Commands);

```

When I see the errors I get, I suppose this is just not supported, right ?

If no, it's fine, I was just curious on trying to be smart for once 😄

**Comment by HomeSpan** [comment]
I can see a number of C++ errors (unrelated to HomeSpan).  First, all arrays in C and C++ start at 0, not 1.  Also, in your for-loop you need to use a ==, not =, for comparisons.  

You are also trying to set a String variable to Service, instead of having the variable be a pointer to a Service. 

I highly recommend taking an online C++ class or two as you start to develop more complex sketches so you can focus more on the HomeSpan aspects of the sketch.

**Reply by Glujaz** [reply]
Yeah, i'm doing this error to often... my Bad. And about the start at 0, I tried that as well initially. But as I did only code on Basic or VBA, I could start from 1, until inputCount included. Seems to be a bit different.

But this..... :

"I highly recommend taking an online C++ class or two"

That hurts ! 😵‍💫 
I discussed on that with a friend yesterday on that topic who codes on C (im just a mechanical engineer), and confirmed your saying ! We are going to do a coding session next week-end together, and explain to me all these pointers and memory thing.

for now, the code I currently have is already enough for my needs, and I was not planning to do more, but I'm curious for at least improving my code on some of my free time 😄 

yeah, my goals was just to be able to turn on the amp when I play music on my AirPort Express, and maybe try another project with an IKEA purifier, but that was pretty much it.

just finishing with : thank you so much for your library. It is so helpful and once you understand some stuff, with some copy and paste, and with chat gpt, you can almost get what you want. And I just learned so much ! So, really, thank you !

**Comment by HomeSpan** [comment]
I'm glad you are enjoying the library and it is a good framework from which to further explore C/C++ since instead of dealing with abstract learning, you can put things into practical use.

C++ is a terrific language but the syntax and some of the concepts can be a bit daunting, especially when first learning.  One of the strengths of C++ is also its largest challenge - there are many different ways of accomplishing the same tasks.  This is good in the sense of providing flexibility, but not so good when trying to learn.  Even among the various HomeSpan examples you will see I used different styles.

For example, sometimes you will see an example where I created pointers to Characteristics, other times I create them without pointers.  The reason is that in some of my first examples, the initial values for the Characteristics were going to be set based on variable inputs when they were first created.  This is best accomplished with using pointers and `new`.  But in other examples I had Characteristics that did not need to be created with anything except the default values, so I simply created them without using pointers and `new`.  And in other examples I mix/match, sometimes even using pointers when they are not strictly necessary.

C++ also has many dozens of libraries that have been under continuous augmentation for 40+ years.  Not even professional C++ developers likely use all aspects of the C++ libraries, since many of them have overlapping functions.  Once you identify the few that make sense for your needs, you tend to use those and ignore the others, though sometimes you may find that another library or way of doing something gives you just the option you needed to simplify your code.  Eventually you will develop your own style and adapt it as needed.

Most important thing is to have fun --- programming in C++ can be frustrating, but rewarding if you stick with it.  Good luck!

**Reply by Glujaz** [reply]
Yeah, I think you actually lost me here with the different examples of pointers.


I think though, what's missing in your page, is a very basic tutorial, like an example where when an update happens (like if a switch is turned on), the ESP32 does xxxx (not just turning a pin on), and does something completely different when the switch is turned off. This would at least help for people who are really starting from scratch, and manage to do a basic code already, so that they have a good example of how to integrate their code they badly built with few knowledge, into HomeSpan.

like having a basic template where you can add what they want inside. I'll try to find some time in the coming weeks to propose something for you.
I had a lot of difficulties to understand homespan, when I could manage something quite basic that the esp32 could do for me already :)

**Reply by HomeSpan** [reply]
Those examples are already there - see the overview page and tutorials 1-5.

**Reply by Glujaz** [reply]
You'll understand once I do it :)
